### ISSUE TITLE
onContextAvailable Not Called When Restoring from Saved Instance State

### ISSUE THREAD

**Issue by 0xcaff** [bluelinelabs/Conductor#524]
`onContextAvailable` isn't called on all controllers on the backstack of ControllerHostedRouter when restoring from saved instance state. It is only called on the first and last.

## Minimal Reproduction

In the context of the demo app.

```java
class NestedRouterController extends BaseController {
    @BindView(R.id.child_content_1)
    FrameLayout childView;

    @Override
    protected void onContextAvailable(@NonNull Context context) {
        super.onContextAvailable(context);
    }

    @Override
    protected View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        return inflater.inflate(R.layout.controller_parent, container, false);
    }

    @Override
    protected void onViewBound(@NonNull View view) {
        Router router = getChildRouter(childView);
        router.pushController(RouterTransaction.with(new ChildController()));
    }

    static class ChildController extends BaseController {
        private static final String KEY = "KEY";

        ChildController() {
        }

        ChildController(Bundle args) {
            super(args);
        }

        @BindView(R.id.button)
        Button button;

        @Override
        protected void onContextAvailable(@NonNull Context context) {
            super.onContextAvailable(context);
        }

        @Override
        protected View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
            return inflater.inflate(R.layout.controller_child, container, false);
        }

        int getId() {
            return getArgs().getInt(KEY, 0);
        }

        @Override
        protected void onViewBound(@NonNull View view) {
            button.setOnClickListener((v) -> {
                Bundle next = new Bundle();
                next.putInt(KEY, getId() + 1);
                getRouter().pushController(RouterTransaction.with(new ChildController(next)));
            });
        }
    }
}
```

Route to `NestedRouterController`. Push the button twice. Attach breakpoints to all `onContextAvailable` methods. Cause the activity to be destroyed and re-created from saved instance state. This can be easily done by enabling don't keep activities in Settings > Developer Options > Apps > Don't Keep Activities, then backgrounding and foregrounding the app.

Observe that onContextAvailable is only called for NestedRouterController and ChildController instances 0 and 2. The inner backstack also has all three controllers on it.

Expected Behavior: onContextAvailable is called for all instances of ChildController including instance 1.

## Why?

I want to use a target controller to communicate with the controller above the current controller in the back stack, without popping the top controller off the backstack. I use dagger to inject dependencies into my controller from DI graph which the activity has a reference to. Currently, injection happens in onCreateView in my app. This works most of the time, except in the case the app is destroyed and re-created from saved instance state and a target controller is used. The controllers are created on the backstack, but when I call a method on the target controller, dependency injection hasn't happened yet. This is because `onCreateView` hasn't been called yet. I tried moving to `onContextAvailable` as the point for injection in my app, but isn't always called when restoring from saved instance state, causing calls from my controller to the target controller to throw because fields are null and haven't been injected yet.

**Comment by EricKuck** [comment]
Thank you for this excellent writeup! This was very easy to reproduce with the given steps and has been fixed in f72b5bed650cfda027a7ee749c19ce00f2aa48fb.

**Comment by PaulWoitaschek** [comment]
@EricKuck Can we release a hotfix for this?

This has a few more implications on the lifecycle. If you use [LifecycleController](conductor-modules/arch-components-lifecycle/src/main/java/com/bluelinelabs/conductor/archlifecycle/LifecycleController.java) with a child router, then [preDestroy](https://github.com/bluelinelabs/Conductor/blob/94c9121bb16f93b481954513a8e3905846829fb2/conductor-modules/arch-components-lifecycle/src/main/java/com/bluelinelabs/conductor/archlifecycle/ControllerLifecycleOwner.java#L54) will throw an `IllegalArgumentException`  because it calls `downEvent` in the `LifecycleRegistry` while in initialized state:
https://github.com/aosp-mirror/platform_frameworks_support/blob/androidx-master-dev/lifecycle/runtime/src/main/java/androidx/lifecycle/LifecycleRegistry.java#L263

```
java.lang.IllegalArgumentException
        at androidx.lifecycle.LifecycleRegistry.downEvent(LifecycleRegistry.java:256)
        at androidx.lifecycle.LifecycleRegistry.backwardPass(LifecycleRegistry.java:307)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:328)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:138)```
